# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/libwebrtcglue/AudioConduit.h
# Commit: 877a0fcc6da9
# Full Hash: 877a0fcc6da9fde4a783cfe30d0aa0b260427ac3
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 9
# Description:
#   Bug 1654112 - Merge conduit controllers into conduits, refactor away setters, and make unittests async. r=bwc,ng
#   
#   The main point of this patch is to harden how conduits handle configuration
#   changes. With conduit controller classes acting in a layer on top of conduits
#   (and their old sync style APIs), for instance
# ==============================================================================

diff -r 970ed33359b5 -r 877a0fcc6da9 dom/media/webrtc/libwebrtcglue/AudioConduit.h
--- a/dom/media/webrtc/libwebrtcglue/AudioConduit.h	Tue Aug 10 14:13:49 2021 +0200
+++ b/dom/media/webrtc/libwebrtcglue/AudioConduit.h	Mon Aug 16 15:57:10 2021 +0200
@@ -7,6 +7,7 @@
 
 #include "mozilla/Attributes.h"
 #include "mozilla/ReentrantMonitor.h"
+#include "mozilla/StateMirroring.h"
 #include "mozilla/TimeStamp.h"
 
 #include "MediaConduitInterface.h"
@@ -19,6 +20,8 @@
  */
 namespace mozilla {
 
+struct DtmfEvent;
+
 /**
  * Concrete class for Audio session. Hooks up
  *  - media-source and target to external transport
@@ -47,45 +50,12 @@
   Maybe<DOMHighResTimeStamp> LastRtcpReceived() const override;
   DOMHighResTimeStamp GetNow() const override;
 
-  MediaConduitErrorCode StopTransmitting() override;
-  MediaConduitErrorCode StartTransmitting() override;
-  MediaConduitErrorCode StopReceiving() override;
-  MediaConduitErrorCode StartReceiving() override;
-
   MediaConduitErrorCode StopTransmittingLocked();
   MediaConduitErrorCode StartTransmittingLocked();
   MediaConduitErrorCode StopReceivingLocked();
   MediaConduitErrorCode StartReceivingLocked();
 
   /**
-   * Function to configure send codec for the audio session
-   * @param sendSessionConfig: CodecConfiguration
-   * @result: On Success, the audio engine is configured with passed in codec
-   *                      for send.
-   *          On failure, audio engine transmit functionality is disabled.
-   * NOTE: This API can be invoked multiple times. Invoking this API may involve
-   *       restarting transmission sub-system on the engine.
-   */
-  MediaConduitErrorCode ConfigureSendMediaCodec(
-      const AudioCodecConfig& codecConfig) override;
-  /**
-   * Function to configure list of receive codecs for the audio session
-   * @param sendSessionConfig: CodecConfiguration
-   * @result: On Success, the audio engine is configured with passed in codec
-   *                      for receive. Also the playout is enabled.
-   *          On failure, audio engine transmit functionality is disabled.
-   *
-   * NOTE: This API can be invoked multiple times. Invoking this API may involve
-   *       restarting transmission sub-system on the engine.
-   */
-  MediaConduitErrorCode ConfigureRecvMediaCodecs(
-      const std::vector<AudioCodecConfig>& codecConfigList) override;
-
-  MediaConduitErrorCode SetLocalRTPExtensions(
-      MediaSessionConduitLocalDirection aDirection,
-      const RtpExtList& extensions) override;
-
-  /**
    * Register External Transport to this Conduit. RTP and RTCP frames from the
    * VoiceEngine shall be passed to the registered transport for transporting
    * externally.
@@ -148,29 +118,36 @@
 
   virtual ~WebrtcAudioConduit();
 
-  /**
-   * Set Local SSRC list.
-   *
-   * This list should contain only a single ssrc.
-   */
-  bool SetLocalSSRCs(const std::vector<uint32_t>& aSSRCs,
-                     const std::vector<uint32_t>& aRtxSSRCs) override;
+  // Call thread.
+  void InitControl(AudioConduitControlInterface* aControl) override;
+
+  // Handle a DTMF event from mControl.mOnDtmfEventListener.
+  void OnDtmfEvent(const DtmfEvent& aEvent);
+
+  // Called when a parameter in mControl has changed. Call thread.
+  void OnControlConfigChange();
+
   std::vector<uint32_t> GetLocalSSRCs() const override;
-  bool SetRemoteSSRC(uint32_t ssrc, uint32_t rtxSsrc) override;
-  bool UnsetRemoteSSRC(uint32_t ssrc) override { return true; }
+
+ private:
+  /**
+   * Override the remote ssrc configured on mRecvStreamConfig.
+   *
+   * Recreates and restarts the recv stream if needed. The overriden value is
+   * overwritten the next time the mControl.mRemoteSsrc mirror changes value.
+   *
+   * Call thread only.
+   */
+  bool OverrideRemoteSSRC(uint32_t ssrc);
+
+ public:
+  void UnsetRemoteSSRC(uint32_t ssrc) override {}
   bool GetRemoteSSRC(uint32_t* ssrc) const override;
-  bool SetLocalCNAME(const char* cname) override;
-  bool SetLocalMID(const std::string& mid) override;
-
-  void SetSyncGroup(const std::string& group) override;
 
   Maybe<webrtc::AudioReceiveStream::Stats> GetReceiverStats() const override;
   Maybe<webrtc::AudioSendStream::Stats> GetSenderStats() const override;
   webrtc::Call::Stats GetCallStats() const override;
 
-  bool InsertDTMFTone(unsigned char payloadType, int payloadFrequency,
-                      int eventCode, int lengthMs) override;
-
   bool IsSamplingFreqSupported(int freq) const override;
 
  private:
@@ -185,17 +162,22 @@
   unsigned int GetNum10msSamplesForFrequency(int samplingFreqHz) const;
 
   // Checks the codec to be applied
-  MediaConduitErrorCode ValidateCodecConfig(const AudioCodecConfig& codecInfo,
-                                            bool send);
+  static MediaConduitErrorCode ValidateCodecConfig(
+      const AudioCodecConfig& codecInfo, bool send);
+  /**
+   * Of all extensions in aExtensions, returns a list of supported extensions.
+   */
+  static RtpExtList FilterExtensions(
+      MediaSessionConduitLocalDirection aDirection,
+      const RtpExtList& aExtensions);
+  static webrtc::SdpAudioFormat CodecConfigToLibwebrtcFormat(
+      const AudioCodecConfig& aConfig);
 
   MediaConduitErrorCode CreateSendStream();
   void DeleteSendStream();
   MediaConduitErrorCode CreateRecvStream();
   void DeleteRecvStream();
 
-  bool RecreateSendStreamIfExists();
-  bool RecreateRecvStreamIfExists();
-
   mozilla::ReentrantMonitor mTransportMonitor;
 
   // Accessed on any thread under mTransportMonitor.
@@ -248,6 +230,38 @@
   // Socket transport service thread. Any thread.
   const nsCOMPtr<nsISerialEventTarget> mStsThread;
 
+  struct Control {
+    // Mirrors and events that map to AudioConduitControlInterface for control.
+    // Call thread only.
+    Mirror<bool> mReceiving;
+    Mirror<bool> mTransmitting;
+    Mirror<Ssrcs> mLocalSsrcs;
+    Mirror<std::string> mLocalCname;
+    Mirror<std::string> mLocalMid;
+    Mirror<Ssrc> mRemoteSsrc;
+    Mirror<std::string> mSyncGroup;
+    Mirror<RtpExtList> mLocalRecvRtpExtensions;
+    Mirror<RtpExtList> mLocalSendRtpExtensions;
+    Mirror<Maybe<AudioCodecConfig>> mSendCodec;
+    Mirror<std::vector<AudioCodecConfig>> mRecvCodecs;
+    MediaEventListener mOnDtmfEventListener;
+
+    // For caching mRemoteSsrc, since another caller may change the remote ssrc
+    // in the stream config directly.
+    Ssrc mConfiguredRemoteSsrc = 0;
+    // For tracking changes to mSendCodec.
+    Maybe<AudioCodecConfig> mConfiguredSendCodec;
+    // For tracking changes to mRecvCodecs.
+    std::vector<AudioCodecConfig> mConfiguredRecvCodecs;
+
+    Control() = delete;
+    explicit Control(const RefPtr<AbstractThread>& aCallThread);
+  } mControl;
+
+  // WatchManager allowing Mirrors to trigger functions that will update the
+  // webrtc.org configuration.
+  WatchManager<WebrtcAudioConduit> mWatchManager;
+
   // Accessed from mStsThread. Last successfully polled RTT
   Maybe<DOMHighResTimeStamp> mRttSec;
 